### PR TITLE
The qpack encoder uses stream_send_atomic to send instructions

### PR DESCRIPTION
**Pull request by ddragana**

### CODE REVIEW

File: neqo-transport/src/send_stream.rs [hunk 1]
```diff
@@ -1232,4 +1257,35 @@ mod tests {
...
+
+        let mut s = SendStream::new(4.into(), 0, Rc::clone(&flow_mgr), conn_events);
+        s.set_max_stream_data(2);
+
+        // Stream is initially blocked (conn:5, stream:2)
+        // and will not accept atomic write of 3 bytes.
+        assert_eq!(s.send_atomic(b"abc").unwrap(), 0);
+
+        // assert non-atomic write works
+        assert_eq!(s.send(b"abc").unwrap(), 2);
```
Comment (martinthomson) [inline neqo-transport/src/send_stream.rs:1275]: This should still send a blocked frame.  This test doesn't appear to test that these arrive at the right moments.  Can you do that?

File: neqo-transport/src/send_stream.rs [hunk 2]
```diff
@@ -629,6 +649,11 @@ impl SendStream {
             return Ok(0);
         }
 
+        if atomic && (can_send_bytes < buf.len() as u64) {
```
[the remark this answers is not in the document]
Comment (ddragana) [reply neqo-transport/src/send_stream.rs:652]: I will make it more compact, but I will open another issue for making avail() return usize. That will not be some lines of code.

[the remark this answers is not in the document]
Comment (martinthomson) [reply neqo-transport/src/send_stream.rs:652]: Yeah, getting the right types for things is hard.  We should agree on what the principles are.

File: neqo-qpack/src/encoder.rs [hunk 3]
```diff
@@ -211,50 +211,49 @@ impl QPackEncoder {
...
+        qdebug!([self], "insert {:?} {:?}.", name, value);
         self.send(conn)?;
         if self.send_buf.len() != 0 {
             return Err(Error::EncoderStreamBlocked);
         }
 
+        let entry_size =
+            u64::try_from(name.len() + value.len() + ADDITIONAL_TABLE_ENTRY_SIZE).unwrap();
+
+        if !self.table.check_insert_possible(entry_size) {
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:229]: This is another case where usize is probably a better type for the argument to check_insert_possible (I would call it just `insert_possible`).

File: neqo-qpack/src/table.rs [hunk 4]
```diff
@@ -32,7 +34,7 @@ impl DynamicTableEntry {
     }
 
     pub fn size(&self) -> u64 {
```
Comment (martinthomson) [inline neqo-qpack/src/table.rs:36]: sizes should be usize :)

File: neqo-transport/src/connection.rs [hunk 5]
```diff
@@ -3380,10 +3385,14 @@ mod tests {
...
-            assert!(matches!(
-                frames[0],
-                (Frame::Stream { .. }, PNSpace::ApplicationData)
-            ));
+            assert_eq!(
+                frames
+                    .iter()
+                    .filter(|i| matches!(i, (Frame::Stream { .. }, PNSpace::ApplicationData)))
+                    .collect::<Vec<_>>()
+                    .len(),
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:3393]: You can use .count() instead of .collect().len().

**Pull request merged**